packages/navigator/src/browser/navigator-tree.ts

```diff
@@ -30,7 +32,102 @@ export class FileNavigatorTree extends FileTree {
     }
 
     async resolveChildren(parent: CompositeTreeNode): Promise<TreeNode[]> {
+        if (FileStatNode.is(parent)) {
+            let resolvedChildren: TreeNode[] = [];
+            let fileStat;
+            if (this.isVirtualRoot(parent)) {
+                fileStat = await this.resolveVirtualRootFileStat(parent);
+                if (fileStat) {
+                    parent.fileStat = fileStat;
+                }
+            } else {
+                fileStat = await this.resolveFileStat(parent);
+            }
+
+            if (fileStat) {
+                resolvedChildren = await this.toNodes(fileStat, parent);
+            }
+            return this.filter.filter(resolvedChildren);
+        }
         return this.filter.filter(super.resolveChildren(parent));
     }
 
+    protected async toNode(fileStat: FileStat, parent: CompositeTreeNode): Promise<FileNode | DirNode> {
+        const id = this.generateFileTreeNodeId(fileStat.uri, parent);
+        const node = this.getNode(id);
+        if (fileStat.isDirectory) {
+            if (DirNode.is(node)) {
+                node.fileStat = fileStat;
+                return node;
+            }
+            return Object.assign(await super.toNode(fileStat, parent), { id });
+        }
+        if (FileNode.is(node)) {
+            node.fileStat = fileStat;
+            return node;
+        }
+        return Object.assign(await super.toNode(fileStat, parent), { id });
+    }
+
+    protected isVirtualRoot(node: CompositeTreeNode): boolean {
```

Comment:
What is the virtual root? `WorkspaceRoot`? Why don't use `WorkspaceRoot.is`?

Reply:
my bad. it is `WorkspaceRootNode`. I called it virtual root because it is a node that does not exist in the file system. I will change it in the way you suggested.